### PR TITLE
Bug 1845409 - File RFC 0010 proposing handling navigation through middleware

### PR DESCRIPTION
**Pull request by MatthewTighe**
This patch includes an RFC for a new architectural goal of handling navigation through middleware moving forward. Currently, there are two commits. 

1. Old proposal for observing navigation state and reacting accordingly.
2. New proposal for handling navigation through middleware.

The first proposal seemed suboptimal after offline conversations with the team, so I've reworked it accordingly. It's included here for context.

Tentative deadline is Feb 2nd.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1845409

### CODE REVIEW

**Review by jonalmeida — APPROVED** [review]
This seems fair to me!

I reviewed the code in Focus to see how it connects the AppStore to an activity that has a lifetime, and I think that flow makes sense for Fenix too for only navigations that that are coming from `HomeActivity`.

I'm curious to see a prototype of this though and if it plays well with the navgraph library.

**Review by MozillaNoah — DISMISSED** [review]
This all makes sense to me. What would be the next steps? Do you have v0.1 or v1 of what this effort looks like?

**Comment by MatthewTighe** [comment]
I would like to get a small prototype together in the next few days focused on tackling the bug this is filed against. I'll announce more broadly when I get something together

**Comment by MatthewTighe** [comment]
I worked up a basic prototype that can be found at https://github.com/mozilla-mobile/firefox-android/pull/4323

**Comment by MatthewTighe** [comment]
I've updated the proposal based on offline conversations with the team in our weekly sync meeting. I've left the original proposal as a separate commit for now for context.

**Comment by MozillaNoah** [comment]
@MatthewTighe, should this be moved out of draft?

**Comment by MatthewTighe** [comment]
> @MatthewTighe, should this be moved out of draft?

Yeah, I will move this out of draft shortly. Was getting a draft pass from @boek offline. 

> Did the Debug Drawer navigation help inform some of this direction?

The Debug Drawer, Review Checker, and History fragment refactor all helped inform this, yes.

> Also, will the prototype PR be updated at all?

Yes, I can update the prototype PR

**Comment by MatthewTighe** [comment]
@MozillaNoah Instead of dealing with a messy diff to update the previous prototype, I've drafted a [new one](https://github.com/mozilla-mobile/firefox-android/pull/5247)

**Review by MozillaNoah — APPROVED** [review]
Nice job with the updates!

File: docs/rfcs/0010-add-state-based-navigation.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	override fun invoke(
+		context: MiddlewareContext<HistoryState, HistoryAction>,
+		next: (HistoryFragmentAction) -> Unit,
+		action: HistoryFragmentAction,
+	) {
+        next(action)
+        when (action) {
+            is HomeButtonClicked -> getNavController().navigate(R.id.home_fragment)
+			is HistoryGroupClicked -> getNavController().navigate(R.id.history_metadata_fragment)
+        }
```
Comment (rocketsroger) [inline docs/rfcs/0010-add-state-based-navigation.md:50]: Nit: indent missing?

Comment (MatthewTighe) [reply docs/rfcs/0010-add-state-based-navigation.md]: Thanks, apparently Android Studio used actual unexpanded tab characters any time I copied a line with a tab in it. Haven't run into a tabs v spaces problem in a longggggg time 😆

File: docs/rfcs/0010-add-state-based-navigation.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+title: Add State-based navigation
+permalink: /rfc/0010-add-state-based-navigation
+---
+
+* Start date: 2023-10-17
+* RFC PR: [4126](https://github.com/mozilla-mobile/firefox-android/pull/4126)
+
+## Summary
+
+Following the acceptance of [0009 - Remove Interactors and Controllers](0009-remove-interactors-and-controllers), Fenix should have a method of navigation that is tied to the `lib-state` model to  provide a method of handling navigation side-effects that is consistent with architectural goals.
```
Comment (t-p-white) [inline docs/rfcs/0010-add-state-based-navigation.md:12]: > that is consistent with architectural goals.
nit: Is there some doc that you can link here to outline the architectural goals?

File: docs/rfcs/0010-add-state-based-navigation.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        next(action)
+        when (action) {
+            is HistoryItemLongClicked -> displayMenuForItem(action.item)
+            is HomeButtonClicked, HistoryGroupClicked -> Unit
+        }
+    }
+}
+```
+
+### 3. The special case of `openToBrowserAndLoad`
```
Comment (t-p-white) [inline docs/rfcs/0010-add-state-based-navigation.md:87]: 👌

**Comment by t-p-white** [comment]
Looks good, no objections from me 🙂 having a consistent method for handling the various aspects of navigation would certainly be beneficial. Appreciate the demo provided too. My only immediate concern is around the defined use cases and how clear the boundaries of responsibility are e.g handling telemetry and other various side effects, but I don't think it's reasonable for you to anticipate _all_ use cases upfront so happy to move forward

**Review by t-p-white — APPROVED** [review]

**Review by Amejia481 — APPROVED** [review]

File: docs/rfcs/0010-add-state-based-navigation.md (hunk 2 above)
Comment (MatthewTighe) [reply docs/rfcs/0010-add-state-based-navigation.md]: Good call, updated!

Comment (t-p-white) [reply docs/rfcs/0010-add-state-based-navigation.md]: Nice 👌

**Pull request merged**